tests/robots/test_robot.py

```diff
@@ -273,3 +290,77 @@ def test_iter_inverse_kinematics(ur5_with_fake_ik):
     assert len(solutions) == 2
     assert str(solutions[0]) == "Configuration((-1.572, -2.560, 2.196, 2.365, 0.001, 1.137), (0, 0, 0, 0, 0, 0), ('shoulder_pan_joint', 'shoulder_lift_joint', 'elbow_joint', 'wrist_1_joint', 'wrist_2_joint', 'wrist_3_joint'))"
     assert str(solutions[1]) == "Configuration((-2.238, -3.175, 2.174, 4.143, -5.616, -6.283), (0, 0, 0, 0, 0, 0), ('shoulder_pan_joint', 'shoulder_lift_joint', 'elbow_joint', 'wrist_1_joint', 'wrist_2_joint', 'wrist_3_joint'))"
+
+
+def test_attach_tool_without_group(ur5_robot_instance, robot_tool1):
+    robot = ur5_robot_instance
+    tool = robot_tool1
+
+    robot.attach_tool(tool)
+
+    assert len(robot.attached_tools) == 1
+    assert robot.main_group_name in robot.attached_tools
+
+
+def test_attach_tool_with_group(ur5_robot_instance, robot_tool1):
+    robot = ur5_robot_instance
+    tool = robot_tool1
+    group_name = "endeffector"
+
+    robot.attach_tool(tool, group=group_name)
+
+    assert len(robot.attached_tools) == 1
+    assert group_name in robot.attached_tools
+
+
+def test_attach_multiple_tools(ur5_robot_instance, robot_tool1, robot_tool2):
+    robot = ur5_robot_instance
+    tool1 = robot_tool1
+    tool2 = robot_tool2
+    group_name = "endeffector"
+
+    robot.attach_tool(tool1)
+    robot.attach_tool(tool2, group=group_name)
+
+    assert len(robot.attached_tools) == 2
+    assert group_name in robot.attached_tools
+    assert robot.main_group_name in robot.attached_tools
+
+
+def test_detach_tool_without_group(ur5_robot_instance, robot_tool1):
+    robot = ur5_robot_instance
+    tool = robot_tool1
+    test_attach_tool_without_group(robot, tool)
```

Comment:
philosophical question: isn't it better to add code duplication but keep the tests completely indepentent?

Reply:
After staring at these for a long few minutes, trying to remember what I even did there, I think I have to agree.